### ISSUE TITLE
Problem playing mp4 music videos on ubuntu 18.04LTS

### ISSUE THREAD

**Issue by mnissov** [UltraStar-Deluxe/Play#246]
Issue type: Bug report

### Actual behaviour
In sing mode when going over a song with a video file a red error message appears saying "Can't play movie [file:///home/user//...]. This happens with both .avi and .mp4 video formats.

The same song repository works with ultra star deluxe.

### Details
Provide some additional information:

- UltraStar Play version: 0.3.1
- Ubuntu 18.04 LTS

**Comment by basisbit** [comment]
Please add more information, like file format and so on. UltraStar Play only supports file formats which Unity framework can play on your OS, plus mp3 for which we added additional support.

**Comment by mnissov** [comment]
> Please add more information, like file format and so on. UltraStar Play only supports file formats which Unity framework can play on your OS, plus mp3 for which we added additional support.

Edited the original, but restated it happens with .avi and .mp4. Not sure if my case has a working video format, all the videos I have are in these formats.

**Comment by paradigmn** [comment]
According to the Unity [manual](https://docs.unity3d.com/2021.1/Documentation/Manual/VideoSources-FileCompatibility.html) only .ogv .webm and .vp8 videos are supported on linux. I made a couple of tests and the only thing that worked for me was a VP8 encoded video with WebM container. The file had to be converted with [ffmpeg](https://trac.ffmpeg.org/wiki/Encode/VP8).

**Comment by achimmihca** [comment]
This surprises me, but you are right. A recent (Feb 2021) [forum post](https://forum.unity.com/threads/error-importing-mp4-file.711203/)

> At this time we don't support MP4 import on Linux. ... H.264 / H.265 are patent-encumbered. With Mac and Windows this is less of an issue since the platforms themselves cover licensing. With Linux the story is murky since no distro does this.
I like the idea of supporting 3rd party codecs at some point. We don't have the bandwidth right now, though.

So mp4 is not supported on Linux. And it seems that adding this codec via Unity API is not supported either.

This sucks, I expected mp4 to work on every platform. Maybe there is some other C# lib that we could use.
There are C# wrappers for ffmpeg: [StackOverflow](https://stackoverflow.com/questions/6717184/net-ffmpeg-wrapper-for-video-playback), [MediaToolkit.NetCore](https://github.com/mtebenev/MediaToolkit.NetCore).
Not sure if these will work with Unity on Linux. Note that we would need a lib that is compatible with our MIT license.

**Comment by basisbit** [comment]
I'd sugest looking at LibVLC instead of ffmpeg - especially after all those years of experience with ffmpeg changing their API again and again even in minor version updates.

**Comment by Niko72** [comment]
> Issue type: Bug report
> 
> ### Actual behaviour
> In sing mode when going over a song with a video file a red error message appears saying "Can't play movie [file:///home/user//...]. This happens with both .avi and .mp4 video formats.
> 
> The same song repository works with ultra star deluxe.
> 
> ### Details
> Provide some additional information:
> 
> * UltraStar Play version: 0.3.1
> * Ubuntu 18.04 LTS

With WEBM the error message is gone−thanks for solving the problem, but the video is not shown! Only the background picture is shown. The only difference is that I am on Ubuntu 20.04LTS. I have not found any settings to change that, nor within `Settings.json`. The TXT file seems to be ok. Is there a shortcut or something else to activate the music video?

**Comment by Niko72** [comment]
> With WEBM the error message is gone−thanks for solving the problem, but the video is not shown! Only the background picture is shown. The only difference is that I am on Ubuntu 20.04LTS. I have not found any settings to change that, nor within `Settings.json`. The TXT file seems to be ok. Is there a shortcut or something else to activate the music video?

I found a solution:

1. Convert the MP4 file with `ffmpeg -i input.mp4 -c:v libvpx -b:v 1M -c:a libvorbis output.webm`
2. Rename the file to `output.vp8`

**Comment by achimmihca** [comment]
Some explanation for the ffmpeg command:
- `-i` the input file
- `-c:v` codec for video
- `-c:a` codec for audio
- `-b:v` bit rate for video
- and the output file is the last parameter as it seems (without `-parameterName` before it)

Does it work when you use `ffmpeg -i input.mp4 -c:v libvpx -b:v 1M -c:a libvorbis output.vp8` directly?
If yes, then the second step (renaming) would not be needed.

Further, the audio in the video is optional when only used for USPlay. So I think `-c:a libvorbis` might not be needed.
There is probably a parameter for ffmpeg to not include any audio.

**Comment by Niko72** [comment]
> Some explanation for the ffmpeg command:
> 
> * `-i` the input file
> * `-c:v` codec for video
> * `-c:a` codec for audio
> * `-b:v` bit rate for video
> * and the output file is the last parameter as it seems (without `-parameterName` before it)
> 
> Does it work when you use `ffmpeg -i input.mp4 -c:v libvpx -b:v 1M -c:a libvorbis output.vp8` directly? If yes, then the second step (renaming) would not be needed.
> 
> Further, the audio in the video is optional when only used for USPlay. So I think `-c:a libvorbis` might not be needed. There is probably a parameter for ffmpeg to not include any audio.

I get the following error message:
```
[NULL @ 0x560174a22f00] Unable to find a suitable output format for 'output.vp8'
output.vp8: Invalid argument
```

**Comment by achimmihca** [comment]
> Are there any plans to support avi and mp4 files in the future

Not from my side in the near future. I would stay with Unity's supported formats.
Of course, I welcome others to implement more file formats using thrid-party libraries (e.g. ffmpeg, libvlc).

**Comment by achimmihca** [comment]
> I welcome others to implement more file formats using thrid-party libraries

There is a Unity specific lib for VLC support: https://code.videolan.org/videolan/vlc-unity
I am unsure about the license though. The license file says 

> vlc-unity is released under LGPLv2.1 (or later) and is also available
under a commercial license.

It is offered as commercial product on the Unity Asset Store.
Feel free to ask on their discord wether this would be allowed to use in UltraStar Play.

I could also imagine that others have implemented *some* support for other video formats for Linux.
Feel free to aks on the Unity forums.

**Comment by watho** [comment]
It seems to me that this vlc-unity is windows only at the moment. Linux support is only on the roadmap. I did not find any other solutions despite some commercial unity plugins.

**Comment by dipzza** [comment]
I made a bash script to automate the process of making your song folder usable on linux.

Given the path of the karaoke folder, it converts all video files to VP8/webm and updates the .txt files.

I would recommend making a copy of your original folder before using it in case anything goes wrong. It deletes the original videos, and only tries to convert those that are not already webm, so you can use it when you download new songs.
```
#!/bin/bash
if [ "$#" -ne 1 ]; then
    echo "Usage: convertovp8.sh musicfolder"
    exit 1
fi


IFS=$(echo -en "\n\b")
VCODEC=vp8
CONTAINER=webm
SONG_FOLDERS=`ls -d ${1}*/`
for SONG_FOLDER in $SONG_FOLDERS
do
        VIDEO=`ls $SONG_FOLDER | grep -v .txt | grep -v .sco | grep -v .jpeg | grep -v .jpg | grep -v .mp3`
        VID_NAME=`echo $VIDEO | awk -F. '{print $1}'`
        VID_EXTENSION=`echo $VIDEO | awk -F. '{print $NF}'`

        if [ -f "$SONG_FOLDER$VIDEO" ] && [ "$VID_EXTENSION" != "$CONTAINER" ]
        then
                ffmpeg -i ${SONG_FOLDER}${VIDEO} -vcodec ${VCODEC} ${SONG_FOLDER}${VID_NAME}.${CONTAINER}
                rm ${SONG_FOLDER}${VIDEO}
                TXT_FILES=`ls $SONG_FOLDER | grep .txt`
                for TXT_FILE in $TXT_FILES
                do
                        echo "sed -i 's/.${VID_EXTENSION}/.${CONTAINER}/g' \"$SONG_FOLDER$TXT_FILE\"" | sh
                done
        fi
done
```

**Comment by achimmihca** [comment]
relates to #344

**Comment by achimmihca** [comment]
Will be implemented with #362
